8.0 Update: SignalR Config: stateful reconnect (try 2)

wadepickett opened the pull request:
Fixes https://github.com/dotnet/AspNetCore.Docs/issues/30674

Added new section: Configure stateful reconnect

(I had to close an earlier PR of this #30829 and start over since I introduced some merge conflicts that were a bit too messy.)

wadepickett commented:
Also updated per @BrennanConroy request (in previous closed version of this PR) to move the server example to the top example.
"I think this should be listed first. Server must allow stateful reconnect otherwise client settings are ignored.
We also don't need to put it for both clients, if it's at the top that should be obvious 😄"

wadepickett commented:
@BrennanConroy, I closed the earlier PR for this and started over with this clean one to get through a merge conflict and a messed up branch.

Your requested change to move the server hub endpoint example up is in.  I also cleaned up some repeated statements to make it a little more concise.

Does this look good?

BrennanConroy commented on aspnetcore/signalr/configuration.md:
```diff
@@ -325,6 +325,52 @@ var connection = new signalR.HubConnectionBuilder()
+
+Stateful reconnect is available in ASP.NET Core 8.0 and later.
+
+Opt-in to stateful reconnect at both the server hub enpoint and the client:
```
```suggestion
Opt-in to stateful reconnect at both the server hub endpoint and the client:
```

BrennanConroy commented on aspnetcore/signalr/configuration.md:
````diff
@@ -325,6 +325,52 @@ var connection = new signalR.HubConnectionBuilder()
+  {
+      options.AllowStatefulReconnects = true;
+  });
+  ```
````
I forgot that the server can also configure its max buffer size.
For global options:
```csharp
builder.AddSignalR(o => o.StatefulReconnectBufferSize = 1000);
```
or for hub specific options:
```csharp
builder.AddSignalR().AddHubOptions<MyHub>(o => o.StatefulReconnectBufferSize = 1000);
```

wadepickett commented:
@BrennanConroy,

- Added the server global and hub specific `StatefulReconnectBufferSize` examples.
- Cleaned up some very minor items in the doc.  
- Updated the 8.0 What's New doc to mirror all changes and linked back to SingalR Config.

Hopefully, this all looks good. Thanks again for your help!

[Internal review: version 8, stateful reconnect section of SignalR Config](https://review.learn.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-8.0&branch=pr-en-us-30850&tabs=dotnet#configure-stateful-reconnect)

[Internal review: What's New: SignalR: Stateful reconnect](https://review.learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-8.0?view=aspnetcore-8.0&branch=pr-en-us-30850#signalr-stateful-reconnect)

tdykstra commented on aspnetcore/signalr/configuration.md:
```diff
@@ -325,6 +325,66 @@ var connection = new signalR.HubConnectionBuilder()
+
+* Temporarily buffering data on the server and client.
+* Acknowledging messages received (ACK-ing) by both the server and client.
+* Recognizing when a connection is returning and replaying messages that might have been sent while the connection was down.
```
Just an idea -- "up" seems a better counterpart to "down" than "returning". 

```suggestion
* Recognizing when a connection is up and replaying messages that might have been sent while the connection was down.
```

wadepickett replied:
We would be assuming a connection is "up" if it is both receiving and returning, so maybe we don't need to be explicit about returning.  Good point. I'll go with "up".

tdykstra commented on aspnetcore/signalr/configuration.md:
````diff
@@ -325,6 +325,66 @@ var connection = new signalR.HubConnectionBuilder()
+  });
+  ```
+
+  Optionally, the maximimum buffer size in bytes allowed by the server can be set globally or for a specific hub with the `StatefulReconnectBufferSize` option:
````
And if not set explicitly, the maximum buffer size is what? (You answer that for the client but I don't see it here for the server.)

Typo on "maximum"
```suggestion
  Optionally, the maximum buffer size in bytes allowed by the server can be set globally or for a specific hub with the `StatefulReconnectBufferSize` option:
```

wadepickett replied:
Great point, I didn't provide the default here.  Checking the source code it is 100,000 (bytes). I will add that.
    /// <summary>
    /// Gets or sets the maximum bytes to buffer per connection when using stateful reconnect.
    /// </summary>
    internal long StatefulReconnectBufferSize { get; set; } = 100_000;

tdykstra commented on aspnetcore/release-notes/aspnetcore-8.0.md:
You already link to the Configuration article -- you could omit the duplication of configuration details here, just announce the availability of SR and explain what it is, then rely on the link to the Configuration doc for the details about how to enable it. Just a thought.

wadepickett replied:
Thanks! A good point.  I considered not calling out the configuration options themselves and linking instead, but looking over the whole What's new doc, we tend to provide an example of the API or property that is new or changed.  So I decided to keep with what that topic was already doing.

tdykstra reviewed (approved):
Looks good, feel free to not take the suggestions other than the typo fix.
